### PR TITLE
BUG: Crosstab margins ignoring dropna

### PR DESCRIPTION
**Pull request by OXPHOS**
To fix bug #12577 : Crosstab margins ignoring dropna

### CODE REVIEW

**Comment by jreback** [comment]
tests!

**Comment by OXPHOS** [comment]
I did test several groups of data on my Mac and they worked out well.. Can I have some hints where could go wrong? Thanks!

**Comment by jreback** [comment]
you need to add tests to the codebase. ``pandas/tools/tests/test_pivot.py``. see the [docs](http://pandas.pydata.org/pandas-docs/stable/contributing.html)

**Comment by nickeubank** [comment]
@OXPHOS Thanks for the PR!

Contributions to pandas require not only that code be modified, but also that you add tests that ensure the proper behavior. This helps ensure that when other people change code in the future, if they do something that breaks the code you've now written, when they run the test suite (which will include your tests) they'll catch the error. 

The section of the dogs that talks about it is here. http://pandas.pydata.org/pandas-docs/stable/contributing.html#test-driven-development-code-writing 

As a basic test, you can basically write out the case from the issue report. You should probably put around line 938 in `pandas/tools/tests/test_pivot.py`

**Comment by OXPHOS** [comment]
Hi @nickeubank , thanks for your instruction and thanks @jreback !

I modified the codes and added tests.
For modification: truncated data pass to _add_margin if dropna == true in pivot.pivot_table()
For tests: added 3 tests in test_pivot.py. Yet I am not sure about the test style (i.e. class name). 

I tried to add another test case:
`df = DataFrame({'a':[1, np.nan, np.nan, np.nan, np.nan, 2], 'b':[np.nan, 3, 4, 4, 4, np.nan]})
print (crosstab(df.a,df.b, margins=True))`

All intersected entries are NaN. However, this case leads to a general fail (even without my modification) and to be honest I don't how what the expected result should be like. 

Error Information:

>     raise ValueError('No objects to concatenate')
ValueError: No objects to concatenate

I assume we need try/except somewhere else upstream (but I am rushing for lunch..)

File: pandas/tools/pivot.py [hunk 1]
```diff
@@ -149,9 +149,15 @@ def pivot_table(data, values=None, index=None, columns=None, aggfunc='mean',
         table = table.fillna(value=fill_value, downcast='infer')
 
     if margins:
-        table = _add_margins(table, data, values, rows=index,
-                             cols=columns, aggfunc=aggfunc,
-                             margins_name=margins_name)
+        if dropna:
+            data_dropna = data[data.notnull().all(axis = 1)]
+            table = _add_margins(table, data_dropna, values, rows=index,
```
Comment (jreback) [inline pandas/tools/pivot.py]: you don't need to repeat the if, just do:

```
if dropna:
   data = data[......]
```

File: pandas/tools/tests/test_pivot.py [hunk 2]
```diff
@@ -935,6 +935,29 @@ def test_crosstab_no_overlap(self):
         expected = pd.DataFrame()
 
         tm.assert_frame_equal(actual, expected)
+        
+    def test_margin_ignore_dropna_bug(self):
+        # Bug#12577
```
Comment (jreback) [inline pandas/tools/tests/test_pivot.py]: give a small descrption of the problem as well (in a comment)

**Comment by jreback** [comment]
pls add a whatsnew entry

**Comment by nickeubank** [comment]
@OXPHOS I think that the "general fail" you note where no objects intersect is actually behaving fine -- it's throwing an informative exception, and I can't think of another way it could be handled. 

Here are [docs](http://pandas.pydata.org/pandas-docs/stable/contributing.html#documenting-your-code) for adding a "whatsnew" entry -- basically, this is just a notification file that gives people a headsup about changes. I think this will be in version 0.18.1 (@jreback sound right?)

File: pandas/tools/pivot.py (hunk 1 above)
Comment (OXPHOS) [inline pandas/tools/pivot.py]: You mean like below? I felt unsafe changing data values but if you're okay with it I'll update the codes.

`        if dropna:
            data = data[data.notnull().all(axis = 1)]
        table = _add_margins(table, data, values, rows=index,
                             cols=columns, aggfunc=aggfunc,
                             margins_name=margins_name)`

**Comment by jreback** [comment]
yep

File: pandas/tools/pivot.py (hunk 1 above)
Comment (jreback) [inline pandas/tools/pivot.py]: how are you ``changing data values``?

Comment (nickeubank) [inline pandas/tools/pivot.py]: I think she means over-writing `data`. That's fine if all tests clear!

Comment (OXPHOS) [inline pandas/tools/pivot.py]: Got it. Thanks!

**Comment by OXPHOS** [comment]
... I am so sorry I ran flake8 as the first step for the tests but didn't go back to it right before commit or after modification. I know it took a long time to run through the test line. I'll do better this time.

**Comment by OXPHOS** [comment]
Wow I passed! Thanks @nickeubank and @jreback guys for your help! I saw a conflict with base branch now. Anything else I should do? Thanks!

File: pandas/tools/tests/test_pivot.py [hunk 3]
```diff
@@ -936,6 +936,35 @@ def test_crosstab_no_overlap(self):
...
         tm.assert_frame_equal(actual, expected)
 
+    def test_margin_ignore_dropna_bug(self):
+        # GH 12577
+        # pivot_table counts null into margin ('All')
+        # when margins=true and dropna=true
+
+        df = pd.DataFrame({'a': [1, 2, 2, 2, 2, np.nan],
+                           'b': [3, 3, 4, 4, 4, 4]})
+        actual = pd.crosstab(df.a, df.b, margins=True)
```
Comment (jreback) [inline pandas/tools/tests/test_pivot.py]: explictly add ``dropna=True`` for these. tests

then add a test for ``dropna=False`` (same set of tests)
and you will find it raises. the ``level`` is ``None`` fails.

```
In [10]: pd.crosstab(df.a,df.b,margins=True,dropna=False)
KeyError: 'Level None not found'
```

**Comment by jreback** [comment]
looks good. see my comment above. Have to handle another case here.

**Comment by OXPHOS** [comment]
Okay so here's what I found:
- This is an independent issue. Apparently no previous tests have conditions `margins=True` and `dropna=False` at the same time.
- For now, I added `if` to `table.columns.names level` in `_add_margins` and avoided the error. I am not sure whether we need the same condition in `table.rows.names level`.
- But the problem raised from the `if not dropna: table = table.reindex_axis(m, axis=1)` in `pivot_table`, which erased all column names. I have been confused by the column names for a while. I was not sure whether the column names would show up on the side. Then it turned out that the show-up of column names completely dependents on `dropna`, which is kinda weird. Apparently this is a problem for a while and I think it's worth fixing. However, I tried to play with `table = table.reindex_axis(m, axis=1)` but failed. I haven't come up with a good idea to solve the problem so far.

**Comment by jreback** [comment]
@OXPHOS ok let's do this.

going to merge your current PR.

Pls open an issue for the other case with a simple repro example (and cross reference to this PR).

**Comment by OXPHOS** [comment]
Will open soon.

**Pull request closed without merging**